tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,191 @@
+//! The validate_repo module owns the 'validate-ami' subcommand and provides methods for validating
+//! a given TUF repository by attempting to load the repository and download its targets.
+
+use super::RepoTransport;
+use crate::repo::{error as repo_error, repo_urls};
+use crate::Args;
+use futures::future::join_all;
+use log::{info, trace};
+use pubsys_config::InfraConfig;
+use rand::seq::SliceRandom;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::fs::File;
+use std::io;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use tempfile::tempdir;
+use tough::{ExpirationEnforcement, Limits, Repository, Settings};
+use url::Url;
+
+/// Validates a set of TUF repositories
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct ValidateRepoArgs {
+    #[structopt(long)]
+    /// Use this named repo infrastructure from Infra.toml
+    repo: String,
+
+    #[structopt(long)]
+    /// The architecture of the repo being validated
+    arch: String,
+    #[structopt(long)]
+    /// The variant of the repo being validated
+    variant: String,
+
+    #[structopt(long, parse(from_os_str))]
+    /// Path to root.json for this repo
+    root_role_path: PathBuf,
+
+    #[structopt(default_value = "100", long)]
+    /// Randomly samples and retrieves specified percentage of targets
+    percent_target_files: u8,
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them.
+async fn retrieve_percentage_of_targets<T: 'static>(
+    repo: &Repository<'_, T>,
+    percentage: u8,
+) -> Result<(), Error>
+where
+    T: tough::Transport,
+    <T as tough::Transport>::Stream: std::marker::Send,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+
+    let mut tasks = Vec::new();
+    for target in targets
+        .keys()
+        .collect::<Vec<&String>>()
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+    {
+        let target = target.to_string();
+        let mut reader = repo
+            .read_target(&target)
+            .context(repo_error::ReadTarget {
+                target: target.to_string(),
+            })?
+            .context(error::TargetMissing {
+                target: target.to_string(),
+            })?;
+        info!("Downloading target: {}", target);
+        tasks.push(tokio::spawn(async move {
+            // tough's `Read` implementation validates the target as it's being downloaded
+            io::copy(&mut reader, &mut io::sink()).context(error::TargetDownload)
```

Comment:
I think we can include `target` in the error context since we have ownership here, right?